File: Data-Structures-and-Algorithms-using-Swift/AlgorithmsChapters/DP/DiceThrow.swift
import Foundation
func findWays(_ n : Int, _ m : Int, _ V : Int) -> Int {
	var dp : [[Int]] = Array(repeating: Array(repeating: 0, count: V + 1), count: n + 1)
	var j : Int = 1, i : Int, k : Int 
	// Table entries for only one dice.
	while (j <= m && j <= V) {
		dp[1][j] = 1
		j += 1
	}

	// i is number of dice, j is Value, k value of dice.
	i = 2
	while (i <= n) {
		j = 1
		while (j <= V) {
			k = 1
			while (k <= j && k <= m) {
				dp[i][j] += dp[i - 1][j - k]
				k += 1
			}
			j += 1
		}
		i += 1
	}
	return dp[n][V]
}

// Testing code.
print(findWays(3,6,6))

// 10

File: Data-Structures-Algorithms-In-TS/AlgorithmsChapters/DP/DiceThrow.ts
function findWays(n: number, m: number, V: number): number {
  const dp: number[][] = new Array(n + 1).fill(0).map(() => new Array(V + 1).fill(0));

  // Table entries for only one dice.
  for (let j = 1; j <= m && j <= V; j++) {
    dp[1][j] = 1;
  }

  // i is number of dice, j is Value, k value of dice.
  for (let i = 2; i <= n; i++) {
    for (let j = 1; j <= V; j++) {
      for (let k = 1; k <= j && k <= m; k++) {
        dp[i][j] += dp[i - 1][j - k];
      }
    }
  }
  return dp[n][V];
}

console.log(`${findWays(3, 6, 6)} `);


/*
10 
*/